src/chain/shared_state.rs

```diff
@@ -242,63 +240,63 @@ impl SharedState {
     /// occurred in the meantime, the keys for sections covering the rest of the address space are
     /// initialised to the old key that was stored for their common ancestor
     /// NOTE: the function as it is currently is not merge-safe.
-    pub fn update_their_keys(
-        &mut self,
-        prefix: Prefix<XorName>,
-        key: BlsPublicKey,
-        our_id: &PublicId,
-    ) {
-        if let Some(&pfx) = self
+    pub fn update_their_keys(&mut self, key_info: TheirKeyInfo) {
+        if let Some((&old_pfx, old_version)) = self
             .their_keys
-            .keys()
-            .find(|pfx| pfx.is_compatible(&prefix))
+            .iter()
+            .find(|(pfx, _)| pfx.is_compatible(&key_info.prefix))
+            .map(|(pfx, info)| (pfx, info.version))
         {
-            let old_key = unwrap!(self.their_keys.remove(&pfx));
-            self.their_recent_keys.push_front((pfx, old_key.clone()));
+            if old_version >= key_info.version || old_pfx.is_extension_of(&key_info.prefix) {
```

Comment:
I thought it was suggested that currently any `merge` related code can be skipped ?
so, maybe the extension check here, and the words in the PR description of `in case of merge`, are not necessary?

Reply:
This is not to handle merge, it is to handle prefix before a split coming after the split. (i.e like in the new test case that would fail without it).